extensions/amp-analytics/0.1/requests.js

```diff
@@ -417,3 +346,121 @@ export class RequestHandler {
     }, interval);
   }
 }
+
+/**
+ * Function that handler extraUrlParams from config and trigger.
+ * @param {!./variables.VariableService} variableService
+ * @param {?JsonObject} configParams
+ * @param {?JsonObject} triggerParams
+ * @param {!./variables.ExpansionOptions} expansionOption
+ * @return {!Promise<!JsonObject>}
+ * @private
+ */
+function getExtraUrlParams(
+  variableService, configParams, triggerParams, expansionOption) {
+  const requestPromises = [];
+  const params = map();
+  // Don't encode param values here,
+  // as we'll do it later in the getExtraUrlParamsString call.
+  const option = new ExpansionOptions(
+      expansionOption.vars,
+      expansionOption.iterations,
+      true /* noEncode */);
+  // Add any given extraUrlParams as query string param
+  if (configParams || triggerParams) {
+    Object.assign(params, configParams, triggerParams);
+    for (const k in params) {
+      if (typeof params[k] == 'string') {
+        requestPromises.push(
+            variableService.expandTemplate(params[k], option)
+                .then(value => { params[k] = value; }));
+      }
+    }
+  }
+  return Promise.all(requestPromises).then(() => {
+    return params;
+  });
+}
+
+/**
+ * Handle the params map and form the final extraUrlParams string
+ * @param {!./variables.VariableService} variableService
+ * @param {!Object} params
+ * @return {string}
+ */
+function getExtraUrlParamsString(variableService, params) {
+  const s = [];
+  for (const k in params) {
+    const v = params[k];
+    if (v == null) {
+      continue;
+    } else {
+      const sv = variableService.encodeVars(k, v);
+      s.push(`${encodeURIComponent(k)}=${sv}`);
+    }
+  }
+  return s.join('&');
+}
+
+/**
+ * Construct the final requestUrl with baseUrl and extraUrlParams
+ * @param {string} baseUrl
+ * @param {!Array<!Promise<string>>} extraUrlParamStrsPromise
+ */
+function constructExtraUrlParamStrs(baseUrl, extraUrlParamStrsPromise) {
+  return Promise.all(extraUrlParamStrsPromise).then(paramStrs => {
+    filterSplice(paramStrs, item => {return !!item;});
+    const extraUrlParamsStr = paramStrs.join('&');
+    let requestUrl;
+    if (baseUrl.indexOf('${extraUrlParams}') >= 0) {
+      requestUrl = baseUrl.replace('${extraUrlParams}', extraUrlParamsStr);
+    } else {
+      requestUrl = appendEncodedParamStringToUrl(baseUrl, extraUrlParamsStr);
+    }
+    return requestUrl;
+  });
+}
+
+/**
+ * Expand the postMessage string
+ * @param {!AMP.BaseElement} baseInstance
+ * @param {string} msg
+ * @param {?JsonObject} configParams
+ * @param {?JsonObject} triggerParams
+ * @param {!./variables.ExpansionOptions} expansionOption
+ * @param {!Object<string, *>} dynamicBindings A mapping of variables to
+ *     stringable values. For example, values could be strings, functions that
+ *     return strings, promises, etc.
+ * @return {Promise<string>}
+ */
+export function expandPostMessage(baseInstance, msg,
```

Comment:
are we able to just reuse the expandTemplate method?

Reply:
Sure we can also choose to expose the `expandTemplate`  `getExtraUrlParams` `getExtraUrlParamsString` and `constructExtraUrlParamStrs` to amp-analytics.js
But I feel it's cleaner this way. Let me know if you have strong opinion.